### PR TITLE
rbd: use clusterID from the volumeContext when handling nodestage request for migration

### PR DESCRIPTION
**Pull request by humblec**
previously we were retriving clusterID using the monitors field
in the volume context at node stage code path. however it is possible to
retrieve clusterID from the volumeContext passed with  nodeStage request

Additional Info:
https://github.com/ceph/ceph-csi/pull/2523#issuecomment-931130464
    
Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

### CODE REVIEW

File: internal/rbd/nodeserver.go
Comment (nixpanic) [inline internal/rbd/nodeserver.go]: just use `failed to parse the migration handle: %v` as error format, the last component after `:` is expected to be the error message

File: internal/rbd/migration_test.go
Comment (nixpanic) [inline internal/rbd/migration_test.go]: I really hate that we're doing it this way. It is very ugly to overload a `tt` variable like this. Please don't introduce these kind of things for new tests, use a different variable instead.

Comment (humblec) [reply internal/rbd/migration_test.go]: well, its just a variable name' ? as long as "memory" is concerned it is totally different location :). No hard opinion to change the variable name though. can be changed.

File: internal/rbd/nodeserver.go
Comment (humblec) [reply internal/rbd/nodeserver.go]: done

File: internal/rbd/nodeserver.go
Comment (nixpanic) [inline internal/rbd/nodeserver.go]: drop the trailing space at `%v "`

**Review by nixpanic — COMMENTED** [review]
I don't have any concerns with this change. It is not clear to me why this is better or what bug it addresses. That could be explained a little more in the commit message.

File: internal/rbd/nodeserver.go
Comment (humblec) [reply internal/rbd/nodeserver.go]: done.

**Comment by humblec** [comment]
> I don't have any concerns with this change. It is not clear to me why this is better or what bug it addresses. That could be explained a little more in the commit message.

my bad, this is actually a small change which marked as a followup item https://github.com/ceph/ceph-csi/pull/2523#issuecomment-931130464.. Added this information in the PR description too. sorry for missing it in first place @nixpanic

**Review by nixpanic — DISMISSED** [review]

**Comment by humblec** [comment]
@Mergifyio rebase

**Review by yati1998 — DISMISSED** [review]

**Comment by humblec** [comment]
@Mergifyio rebase

**Review by Madhu-1 — DISMISSED** [review]
Please remove GetClusterIDFromMon and   readClusterInfoWithMon which is not required anymore https://github.com/ceph/ceph-csi/pull/2512/files#diff-b9c67ae3efee5ec6b2d1889405ffaf967a4808f3d9a12a287ac6faa9ba19d762R169-R207 .
Please also make required changes in E2E to remove monitor information from PV.CSI  object as  its not required anymore.

File: internal/rbd/nodeserver.go
Comment (Madhu-1) [inline internal/rbd/nodeserver.go]: `failed to parse the migration handle`  to `failed to parse volumeID`

File: internal/rbd/nodeserver.go
Comment (Madhu-1) [inline internal/rbd/nodeserver.go]: as you already know the clusterID can't we just add that to volumeContext in kubernetes itself?

**Review by Madhu-1 — COMMENTED** [review]
@humblec sorry for  it late, i have a question PTAL

**Comment by humblec** [comment]
> Please remove GetClusterIDFromMon and readClusterInfoWithMon which is not required anymore https://github.com/ceph/ceph-csi/pull/2512/files#diff-b9c67ae3efee5ec6b2d1889405ffaf967a4808f3d9a12a287ac6faa9ba19d762R169-R207 . Please also make required changes in E2E to remove monitor information from PV.CSI object as its not required anymore.

@Madhu-1 I kept that out there in case if we want to use the same in future , but I can remove that as well.  Done.

File: internal/rbd/nodeserver.go
Comment (humblec) [reply internal/rbd/nodeserver.go]: @Madhu-1 its possible on both ways, however kept this way as volID considered to be CSI  handler which can be depended for the operations.

Comment (Madhu-1) [reply internal/rbd/nodeserver.go]: in nodeStage operation, we retrieve the clusterID from the context. IMO we should take the same route as it's possible now.

Comment (humblec) [reply internal/rbd/nodeserver.go]: @Madhu-1 if really needed I can do that, wdyt?  why I am in bit hesitation here is, it also need changes in the migration Context func ...etc. Can that be done later? I can mark it as todo for now . Is that fine ?

Comment (Madhu-1) [reply internal/rbd/nodeserver.go]: instead of adding complexity am more inclined to do this as it's going to be minimal change.

Comment (humblec) [reply internal/rbd/nodeserver.go]: well, this patch submission was done to use mons hash as in the previous PR review comment, now the change request is actually different :).

Comment (Madhu-1) [reply internal/rbd/nodeserver.go]: @humblec I do not always think about all the possibilities sometimes we suggesting things and when actually reviewing PR's sometimes I will try to think about the possibilities. if we get good approach IMO not a bad to change in between instead of changing later :)

Comment (humblec) [reply internal/rbd/nodeserver.go]: `@humblec I do not always think about all the possibilities sometimes we suggesting things and when actually reviewing PR's sometimes I will try to think about the possibilities. if we get good approach IMO not a bad to change in between instead of changing later :)` - hmmm..as mentioned both works, any way addressing this comment then.

**Comment by humblec** [comment]
`Error from server: error when creating "https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/v4.0.0/client/config/crd/snapshot.storage.k8s.io_volumesnapshotcontents.yaml": etcdserver: request timed out
script returned exit code 1`

**Comment by humblec** [comment]
I have to adjust the commit message and little bit of splitting the last commit.. on it

**Comment by humblec** [comment]
> I have to adjust the commit message and little bit of splitting the last commit.. on it

Done..

**Comment by humblec** [comment]
@Mergifyio rebase

**Comment by humblec** [comment]
Can you please review this ?

**Review by yati1998 — COMMENTED** [review]
LGTM. Will approve once any of the maintainers approve it.

File: internal/rbd/migration_test.go
Comment (Madhu-1) [inline internal/rbd/migration_test.go]: this test seems redundant as there is no clusteID validation in `isMigrationContext` can you please remove this test?

File: internal/rbd/nodeserver.go
Comment (Madhu-1) [inline internal/rbd/nodeserver.go]: IMO this check can be removed as the clusterID validation will be done when populating the rbdVolume just like other NodeStageRequest, we don't need to tread this as special

**Review by Madhu-1 — COMMENTED** [review]
Currently, we have 3 e2e tests and we can move the duplicate test for Block and See if you can adjust other tests to validate Image deletion.

* validate RBD migration+static Block PVC Deletion
* validate RBD migration+static FileSystem PVC
* validate RBD migration+static Block PVC

File: internal/rbd/nodeserver.go
Comment (humblec) [reply internal/rbd/nodeserver.go]: Thought an extra check here is good for fail early, but removing it.

**Comment by humblec** [comment]
@Madhu-1 addressed the final comments. ptal.. thanks..

File: internal/rbd/migration.go
Comment (Madhu-1) [inline internal/rbd/migration.go]: this is not required anymore and can be removed?

Comment (humblec) [reply internal/rbd/migration.go]: @Madhu-1 done. ptal

**Review by Madhu-1 — APPROVED** [review]

**Review by yati1998 — APPROVED** [review]

**Comment by humblec** [comment]
@Mergifyio rebase

**Pull request merged**
